src/ray/common/memory_monitor.cc

```diff
@@ -111,41 +111,107 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
+  //   Linux OS tries to fully use the whole physical memory size, so that in many
+  //   cases we can observe that the system has very little `MemFree` size,
+  //   but at the same time system might have a large `MemAvailable` size,
+  //   then this part `MemAvailable - MemFree` size is borrowed by Linux OS
+  //   to cache pages / buffers , BUT, once user process requests to allocate memory,
+  //   and there is no sufficient free memory, OS will evict cache data out of this
+  //   part memory and allocate them to user proces.
+  //
+  // Explanation: What's the part of `shmem` and why we should treat it as "in-used" ?
+  //   `shmem` overview: https://man7.org/linux/man-pages/man7/shm_overview.7.html
+  //   Ray object store use `/dev/shm`, which is a `tmpfs` file system,
+  //   In linux, `tmpfs` file system uses `shmem` memory.
+  //   Note that `tmpfs` file system might swap data to disk (when option noswap=False)
+  //   but `shmem` value always means the `tmpfs` data size in physical memory.
+  //   We can read `shmem` value from /proc/meminfo `Shmem` item or from
+  //   /sys/fs/cgroup/memory/memory.stat `total_shmem` item.
+  //
+  // Explanation: Why don't use cgroup
+  //  `memory.stat.total_rss` and `memory.stat.inactive_file_bytes` to compute the
+  // in-used memory ?
+  //   In my test using these values can't calculate out correct used memory number,
+  //   cgroup official doc is fuzzy when describing these values.
+  //   But in my test,
+  //   cgroup `memory.usage_in_bytes` value perfectly matches (`MemTotal` - `MemFree`) value,
+  //   so I am sure that `memory.usage_in_bytes` value is correct.
+  //   and cgroup `memory.stat.total_cache` value also perfectly matches
+  //   (`MemAvailable` - `MemFree`) value,
+  //
+  // Correctness testing criteria:
+  //  - Ensuring the calculated value is consistent with values computed from /proc/meminfo
+  //    Note that cgroup mem file is consistent with /proc/meminfo file unless
+  //    there is bug in cgroup, i.e. we can read MemTotal / MemAvailable / Shmem
+  //    from /proc/meminfo file, then the calculated value by this function
+  //    should equals to `MemTotal` - (`MemAvailable` - `Shmem`)
+  //
+  //  - Prepare an idle OS environment that has a large number of `free` memory.
+  //    Use dd command to write a large file to disk, after dd  completes,
+  //    the calculated used_memory value should keep nearly the same with the value
+  //    before dd execution.
+  //    But note that free memory will decrease significantly because OS
+  //    cached the file data as part of "OS_managed_cache_and_buffer" I mentioned above.
+  //
+  //  - If we use dd command to write a large file to /dev/shm,
+  //    and no swapping occurs (you can use `free -h` to check whether swap size increases),
+  //    after dd completes, the calculated used-memory value should be nearly
+  //    previous_in_used_memory_bytes + bytes_of_written_file
```

Comment:
This part verification is very important.

The current `GetCGroupV1MemoryUsedBytes` just reads some items from CGroup mem stat file without fully understanding their meanings and without verification, and it turns out to be wrong number and causes this hidden bug.